File: packages/b2c-tooling-sdk/src/auth/jwt-utils.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
/**
 * Shared helpers for working with decoded JWT access tokens. Centralizes the
 * `exp` / `scope` extraction patterns that previously lived in multiple auth
 * strategy files.
 *
 * @module auth/jwt-utils
 */
import {decodeJWT} from './oauth.js';

/** Default buffer (seconds) before token `exp` to consider the token expired. */
export const DEFAULT_EXPIRY_BUFFER_SEC = 60;

/**
 * Extract `scope` from a decoded JWT payload. Tolerates both the (legacy)
 * space-delimited string form and the array form.
 */
export function extractJwtScopes(payload: Record<string, unknown>): string[] {
  const scope = payload.scope as string | string[] | undefined;
  if (scope == null) return [];
  return Array.isArray(scope) ? scope : scope.split(' ');
}

/**
 * Decode a JWT and return its `expires` Date and `scopes` array. Errors from
 * `decodeJWT` propagate to the caller — callers that want a soft-fail should
 * wrap this in try/catch.
 */
export function decodeJwtTokenInfo(token: string): {expires: Date; scopes: string[]} {
  const decoded = decodeJWT(token);
  const exp = typeof decoded.payload.exp === 'number' ? decoded.payload.exp : 0;
  return {
    expires: new Date(exp * 1000),
    scopes: extractJwtScopes(decoded.payload as Record<string, unknown>),
  };
}

/**
 * Returns `true` if the token is non-empty, decodes successfully, has not
 * expired (with a small buffer), and includes all required scopes.
 *
 * @param token - The JWT access token
 * @param requiredScopes - Scopes that must all be present in the token (default: none)
 * @param expiryBufferSec - Treat token as expired this many seconds before its real `exp`
 */
export function isJwtTokenValid(
  token: string,
  requiredScopes: string[] = [],
  expiryBufferSec: number = DEFAULT_EXPIRY_BUFFER_SEC,
): boolean {
  if (!token) return false;
  let info: {expires: Date; scopes: string[]};
  try {
    info = decodeJwtTokenInfo(token);
  } catch {
    return false;
  }
  const nowSec = Math.floor(Date.now() / 1000);
  const expSec = Math.floor(info.expires.getTime() / 1000);
  if (expSec === 0 || nowSec >= expSec - expiryBufferSec) return false;
  if (requiredScopes.length > 0 && !requiredScopes.every((s) => info.scopes.includes(s))) {
    return false;
  }
  return true;
}

File: python/src/b2c_tooling_sdk/auth/jwt_utils.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""Helpers for decoding and validating JWT access tokens.

Mirrors ``src/auth/jwt-utils.ts`` (plus ``decodeJWT`` from ``oauth.ts``). Tokens
are decoded **without** signature verification — these are Account Manager
access tokens the SDK itself requested, used only to read ``exp`` / ``scope``.
"""

from __future__ import annotations

import base64
import json
import time
from datetime import datetime, timezone
from typing import Any

from b2c_tooling_sdk.auth.types import DecodedJWT

#: Default buffer (seconds) before token ``exp`` to consider the token expired.
DEFAULT_EXPIRY_BUFFER_SEC = 60


def _b64url_decode(segment: str) -> bytes:
    """Decode a base64url (or plain base64) segment, tolerating missing padding."""
    normalized = segment.replace("-", "+").replace("_", "/")
    padding = "=" * (-len(normalized) % 4)
    return base64.b64decode(normalized + padding)


def decode_jwt(token: str) -> DecodedJWT:
    """Decode a JWT into its header and payload without verifying the signature.

    :raises ValueError: if the token is not a well-formed three-part JWT.
    """
    parts = token.split(".")
    if len(parts) != 3:
        raise ValueError("Invalid JWT format")
    header = json.loads(_b64url_decode(parts[0]))
    payload = json.loads(_b64url_decode(parts[1]))
    return DecodedJWT(header=header, payload=payload)


def extract_jwt_scopes(payload: dict[str, Any]) -> list[str]:
    """Extract ``scope`` from a decoded JWT payload (array or space-delimited string)."""
    scope = payload.get("scope")
    if scope is None:
        return []
    if isinstance(scope, list):
        return [str(s) for s in scope]
    return str(scope).split(" ")


def decode_jwt_token_info(token: str) -> tuple[datetime, list[str]]:
    """Return ``(expires, scopes)`` for a token. Propagates decode errors."""
    decoded = decode_jwt(token)
    exp = decoded.payload.get("exp")
    exp_seconds = exp if isinstance(exp, (int, float)) else 0
    expires = datetime.fromtimestamp(exp_seconds, tz=timezone.utc)
    return expires, extract_jwt_scopes(decoded.payload)


def is_jwt_token_valid(
    token: str,
    required_scopes: list[str] | None = None,
    expiry_buffer_sec: int = DEFAULT_EXPIRY_BUFFER_SEC,
) -> bool:
    """Return ``True`` if the token decodes, is unexpired (with buffer), and has all scopes."""
    if not token:
        return False
    required_scopes = required_scopes or []
    try:
        expires, scopes = decode_jwt_token_info(token)
    except Exception:
        return False
    now_sec = int(time.time())
    exp_sec = int(expires.timestamp())
    if exp_sec == 0 or now_sec >= exp_sec - expiry_buffer_sec:
        return False
    # Empty required_scopes -> all() is vacuously True.
    return all(s in scopes for s in required_scopes)


__all__ = [
    "DEFAULT_EXPIRY_BUFFER_SEC",
    "decode_jwt",
    "extract_jwt_scopes",
    "decode_jwt_token_info",
    "is_jwt_token_valid",
]
